bokehjs/src/lib/models/widgets/tables/data_cube.ts

```diff
@@ -0,0 +1,360 @@
+    const totals: GroupTotals = { avg: {}, max: {}, min: {}, sum: {} }
+    const { source: { data } } = this
+    const keys = Object.keys(data)
+    const items = group.rows.map((i: number) => keys.reduce((o, c) => ({ ...o, [c]: data[c][i] }), {}))
+
+    aggregators.forEach((aggregator: RowAggregator) => {
+      aggregator.init()
+      items.forEach((item: Item) => aggregator.accumulate(item))
+      aggregator.storeResult(totals)
+    })
+    return totals
+  }
+
+  private addTotals(groups: Group[], level = 0): void {
+    const { aggregators, collapsed: groupCollapsed } = this.groupingInfos[level]
+    const toggledGroups = this.toggledGroupsByLevel[level]
+
+    groups.forEach((group: Group) => {
+      if (group.groups) {
+        this.addTotals(group.groups, level + 1)
+      }
+
+      if (aggregators.length && group.rows.length) {
+        group.totals = this.calculateTotals(group, aggregators)
+      }
+
+      group.collapsed = groupCollapsed !== toggledGroups[group.groupingKey]
+      group.title = group.value
+    })
+  }
+
+  private flattenedGroupedRows(groups: Group[], level = 0): (Group | number)[] {
+    const rows: (Group | number)[] = []
+
+    groups.forEach((group: Group) => {
+      rows.push(group)
+      if (!group.collapsed) {
+        const subRows = group.groups
+          ? this.flattenedGroupedRows(group.groups, level + 1)
+          : group.rows
+        rows.push(...subRows)
+      }
+    })
+    return rows
+  }
+
+  refresh(): void {
+    const groups = this.extractGroups(this.view.indices)
+    const labels = this.source.data[this.columns[0].field]
+
+    if (groups.length) {
+      this.addTotals(groups)
+      this.rows = this.flattenedGroupedRows(groups)
+      this.target.data = {
+        row_indices: this.rows.map((value: Group | number) => value instanceof Group ? value.rows : value),
+        labels: this.rows.map((value: Group | number) => value instanceof Group ? value.title : labels[value]),
+      }
+    }
+  }
+
+  getLength(): number {
+    return this.rows.length
+  }
+
+  getItem(i: number): Item {
+    const item = this.rows[i]
+    const { source: { data } } = this
+
+    return item instanceof Group
+      ? item
+      : Object.keys(data)
+          .reduce((o, c) => ({ ...o, [c]: data[c][item] }), { [DTINDEX_NAME]: item })
+  }
+
+  getItemMetadata(i: number): Metadata | null {
+    const myItem = this.rows[i]
+    const { level } = myItem
+    const columns = this.columns.slice(1)
+
+    const aggregators = myItem instanceof Group
+      ? this.groupingInfos[level].aggregators
+      : []
+
+    function adapter(column: Column): Partial<Column> {
+      const { field: myField, formatter } = column
+      const aggregator = aggregators.find(({ field_ }) => field_ === myField)
+
+      if (aggregator) {
+        const { key } = aggregator
+        return {
+          formatter(row: any, cell: any, _value: any, columnDef: any, dataContext: any): string {
```

Comment:
Any instances of `any` type should be either replaced with more specific types (as much as feasible) or with `unknown` type. At some point `any` will be banned from bokehjs' source code.